### ISSUE TITLE
GF comparison broken wth flint

### ISSUE THREAD

**Issue by J08nY** [sympy/sympy#26791]
https://github.com/sympy/sympy/pull/25663 also broke the following code:

```python
> from sympy import FF, Integer
> K = FF(11)
> a = K(3)
> a == 3
  True
> a == -8
  True

# Currently (flint)
> a == Integer(3)
  True
> a == Integer(-8)
  False

# Before
> a == Integer(3)
  True
> a == Integer(-8)
  True
```

A simple wrapping of `Integer` around a pure int breaks comparisons. This is a significant regressionin a rather basic feature.

**Comment by oscarbenjamin** [comment]
You seem to be mixing domain elements and sympy expressions in a way that is not intended or that I would discourage.

Mixing positive characteristic domain elements with ordinary sympy expressions is going to have many problems because ordinary sympy expressions are always assumed to be in characteristic zero.

This is the SymPy 1.12.1 behaviour (and also current behaviour when flint is not used for ground types):
```python
In [1]: FF(11)(3) == Integer(-8)
Out[1]: True

In [2]: Integer(-8) == FF(11)(3)
Out[2]: False
```
This is the SymPy 1.13 behaviour when flint is used:
```python
In [1]: FF(11)(3) == Integer(-8)
Out[1]: False

In [2]: Integer(-8) == FF(11)(3)
Out[2]: False
```

**Comment by J08nY** [comment]
Well, I only tried this after encountering #26789.
I had two lists, one of `FF` elements and one of `Integer` that I was comparing by converting the Integers over to the field (because that is where I wanted to do the comparison). When that stopped working because of #26789 this was the obvious other thing to try.

However, what I am expecting from the behavior is not that wild. Take for example SageMath:
```sage
sage: K = GF(11)
sage: i = Integer(3)
sage: K(i)
3
sage: K(i) == i
True
sage: K(i) == 3
True
sage: K(i) == -8
True
sage: 3 == K(i)
True
sage: -8 == K(i)
True
```

**Comment by oscarbenjamin** [comment]
The python-flint types do behave like that for `int`:
```python
In [3]: from flint import nmod

In [4]: nmod(3, 5) == 3
Out[4]: True

In [5]: nmod(3, 5) == -2
Out[5]: True
```
I think there are two problems here:

1. sympify should reject GF elements (meaning that `Integer.__eq__` would return `NotImplemented`.
2. python-flint's `nmod` etc types should maybe try `__index__` when comparing for equality (so that `nmod.__eq__` would recognise `Integer`).

**Comment by oscarbenjamin** [comment]
> 2\. python-flint's `nmod` etc types should maybe try `__index__` when comparing for equality

This change is needed in python-flint rather than sympy because sympy has no control of `nmod.__eq__`.

Other python-flint types that are used in sympy 1.13 are used only inside wrapper classes that can adapt the interface to preserve compatibility. The overhead of a wrapper class would be too great for `GF` though (python-flint makes GF much faster because it does not have the overhead of Python methods).

**Comment by J08nY** [comment]
Interesting! The no-overhead GF impl is something I am looking forward to quite a bit, although in [my usecase](https://github.com/J08nY/pyecsca/blob/master/pyecsca/ec/mod.py) I likely need a wrapper class (without reworking a bunch of stuff) and thus will not see that much gain.

**Comment by oscarbenjamin** [comment]
I've had a look at the linked code and I am not sure I fully understand it but I wonder if SymPy's GF is really what you want there. I'm not sure what GF is doing that is useful in that code.

**Comment by J08nY** [comment]
> I've had a look at the linked code and I am not sure I fully understand it but I wonder if SymPy's GF is really what you want there. I'm not sure what GF is doing that is useful in that code.

Well the lined code does not do that right now, but other code uses GF and polynomials over it for some root finding etc..
In the linked code I will likely use python-flint directly.

**Comment by oscarbenjamin** [comment]
> In the linked code I will likely use python-flint directly.

I think that makes sense if you are happy to depend on python-flint. The advantage of using SymPy's GF here is just that it provides a pure Python implementation if python-flint is not installed but then transparently speeds up when python-flint is installed.

> other code uses GF and polynomials over it for some root finding etc..

This also could possibly use python-flint directly but again any features python-flint provides for this will be wrapped by SymPy along with a pure Python implementation as fallback.

It seems I overlooked using python-flint for univariate polynomials over GF in sympy because I forgot to merge gh-25940 before releasing SymPy 1.13.

**Comment by J08nY** [comment]
> You seem to be mixing domain elements and sympy expressions in a way that is not intended or that I would discourage.

To get back to this, looking again at my code, all sorts of things are now broken because of this. I assumed having finite field elements in expressions was just fine. How else would one compute with them?

Let's say I have an expression like `4 / c - d  + r`. What I would do is simpify this, stuff some GF(p) values in for `c` and `d`, then get a polynomial in `r` over some GF(p) and do root finding (this is necessary because the general case can be more complicated than this). How would one do this now? Like a quarter of my test suite is failing due to this.
Do I just replace the GF elements with [`Mod`](https://docs.sympy.org/latest/modules/core.html#sympy.core.mod.Mod) when I am operating with the expression and then get rid of them when I have a polynomail I want to root find?

**Comment by oscarbenjamin** [comment]
> Let's say I have an expression like `4 / c - d + r`. What I would do is simpify this, stuff some GF(p) values in for `c` and `d`

I don't understand how you would get modular division like this (assuming that is what is intended by `4/c`). Note that all manipulations of the expression will assume characteristic zero. If you substitute `GF(p)` values they will just become integers.

> Do I just replace the GF elements with [`Mod`](https://docs.sympy.org/latest/modules/core.html#sympy.core.mod.Mod) when I am operating with the expression and then get rid of them when I have a polynomail I want to root find?

Mod would be the way to represent GF integers in a sympy expression although very little symbolic manipulation does anything useful with expressions containing `Mod` and it is not compatible with Poly. Also I don't think that there is a way represent modular inverse.

Generally if you can use polynomial expressions then you can do things like:
```python
In [1]: ground_roots((x - 1)*(x - 2), modulus=11)
Out[1]: {1: 1, 2: 1}

In [2]: Poly((x - 1)*(x - 2), modulus=11)
Out[2]: Poly(x**2 - 3*x + 2, x, modulus=11)

In [3]: Poly((x - 1)*(x - 2), domain=GF(11))
Out[3]: Poly(x**2 - 3*x + 2, x, modulus=11)
```
The usual way that the domain system is used is indirectly via `Poly` or via functions like `ground_roots` that use `Poly` internally.

**Comment by oscarbenjamin** [comment]
Another possibility is that you could use a rational function field like `GF(11)(x,y)`:
```python
In [5]: R, c, d, r = field('c, d, r', domain=GF(11))

In [6]: 4 / c - d + r
Out[6]: (10*c*d + c*r + 4)/(c)
```

**Comment by J08nY** [comment]
> > Let's say I have an expression like `4 / c - d + r`. What I would do is simpify this, stuff some GF(p) values in for `c` and `d`
> 
> I don't understand how you would get modular division like this (assuming that is what is intended by `4/c`). Note that all manipulations of the expression will assume characteristic zero. If you substitute `GF(p)` values they will just become integers.

I am getting the expressions as strings from a source I do not control. They can be things like:
```
Z1 == 1
ZZ2 == 1
c2 == (2 * c)
b4 == (4 * b)
ZZZ1 == 1
a24 == ((a + 2) / 4)
ccd2 == (((2 * c) * c) * d)
cc4 == ((4 * c) * c)
d2 == (2 * d)
b2 == (2 * b)
(k * c) == 1
k == (2 * d)
Z1 == Z2
(i ** 2) == (- 1)
ZZZ2 == 1
r2 == (2 * r)
b3 == (3 * b)
s == ((1 + r) / (1 - r))
X2 == 1
cc2 == ((2 * c) * c)
half == (1 / 2)
ZZ1 == 1
Z2 == 1
c == 1
ccd == ((c * c) * d)
```
The overall task is to, given some GF(p) values for variables contained in those statements, figure out whether the statement holds (if no free variables remain) or if just one free variable remains find its value over GF(p).

Thanks for the help btw. This change is causing me quite a headache as I was relying on something that was apparently very brittle.

**Comment by oscarbenjamin** [comment]
I don't know how you intend division to be interpreted here. I would first try to eliminate all of the denominators to get polynomials:
```python
from sympy import *

eqs_string = """\
Z1 == 1
ZZ2 == 1
c2 == (2 * c)
b4 == (4 * b)
ZZZ1 == 1
a24 == ((a + 2) / 4)
ccd2 == (((2 * c) * c) * d)
cc4 == ((4 * c) * c)
d2 == (2 * d)
b2 == (2 * b)
(k * c) == 1
k == (2 * d)
Z1 == Z2
(i ** 2) == (- 1)
ZZZ2 == 1
r2 == (2 * r)
b3 == (3 * b)
s == ((1 + r) / (1 - r))
X2 == 1
cc2 == ((2 * c) * c)
half == (1 / 2)
ZZ1 == 1
Z2 == 1
c == 1
ccd == ((c * c) * d)
""".splitlines()

def make_eq(s):
    lhs, rhs = s.split(' == ')
    return Eq(parse_expr(lhs), parse_expr(rhs))

equations = [make_eq(s) for s in eqs_string]

eqs_num = []
eqs_den = []
for eq in equations:
    eq = eq.lhs - eq.rhs
    num, den = eq.as_numer_denom()
    eqs_num.append(num)
    eqs_den.append(den)

eqs_poly1 = [eq.as_poly(modulus=11) for eq in eqs_num]
eqs_poly2, opt = parallel_poly_from_expr(eqs_num, modulus=11)

for p in eqs_poly1:
    print(p)
for p in eqs_poly2:
    print(p)
```
Now you can substitute values into the polynomials. You may also want to check if the denominators are zero.
```python
In [26]: print(eqs_poly2[0])
Poly(Z1 - 1, r, r2, s, a, a24, b, b2, b3, b4, c, c2, d, d2, i, k, X2, Z1, Z2, ZZ1, ZZ2, ZZZ1, ZZZ2, cc2, cc4, ccd, ccd2, half, modulus=11)

In [27]: print(eqs_poly2[0].gens)
(r, r2, s, a, a24, b, b2, b3, b4, c, c2, d, d2, i, k, X2, Z1, Z2, ZZ1, ZZ2, ZZZ1, ZZZ2, cc2, cc4, ccd, ccd2, half)

In [28]: vals = [1]*27

In [29]: eqs_poly2[0](*vals)
Out[29]: 0

In [35]: print(eqs_poly2[0].eval(Symbol('Z1'), 1))
Poly(0, r, r2, s, a, a24, b, b2, b3, b4, c, c2, d, d2, i, k, X2, Z2, ZZ1, ZZ2, ZZZ1, ZZZ2, cc2, cc4, ccd, ccd2, half, modulus=11)

In [36]: print(eqs_poly2[0].eval(Symbol('Z1'), -10))
Poly(0, r, r2, s, a, a24, b, b2, b3, b4, c, c2, d, d2, i, k, X2, Z2, ZZ1, ZZ2, ZZZ1, ZZZ2, cc2, cc4, ccd, ccd2, half, modulus=11)

In [41]: print(eqs_poly2[0].subs({Symbol('Z1'): -10}))
Poly(0, r, r2, s, a, a24, b, b2, b3, b4, c, c2, d, d2, i, k, X2, Z2, ZZ1, ZZ2, ZZZ1, ZZZ2, cc2, cc4, ccd, ccd2, half, modulus=11)
```

**Comment by oscarbenjamin** [comment]
> if just one free variable remains find its value over GF(p).

You might find it useful to compute a Groebner basis before or after substituting values for the symbols:
```python
In [42]: gb = groebner(eqs_poly2)

In [43]: for p in gb: print(p.as_expr())
2*r - r2
r2*s + r2 - 2*s + 2
a - 4*a24 + 2
4*b - b4
2*b2 - b4
4*b3 - 3*b4
c - 1
c2 - 2
2*d - 1
d2 - 1
i**2 + 1
k - 1
X2 - 1
Z1 - 1
Z2 - 1
ZZ1 - 1
ZZ2 - 1
ZZZ1 - 1
ZZZ2 - 1
cc2 - 2
cc4 - 4
2*ccd - 1
ccd2 - 1
2*half - 1
```
Note that this could prove the unsatisfiability of the system before substituting any values:
```python
In [52]: groebner(eqs_poly + [Symbol('Z1')-2])
Out[52]: 
GroebnerBasis([1], r, r₂, s, a, a₂₄, b, b₂, b₃, b₄, c, c₂, d, d₂, i, k, X₂, Z₁, Z₂, ZZ₁, ZZ₂, ZZZ₁, ↪

↪  ZZZ₂, cc₂, cc₄, ccd, ccd₂, half, domain=ℤ, order=lex)
```

If you substitute values for all except one symbol then the Groebner basis will be a single polynomial whose roots can be found using `ground_roots`:
```python
In [47]: gb[10].as_expr()
Out[47]: 
 2    
i  + 1

In [48]: gb[10].as_expr().as_poly(modulus=11)
Out[48]: Poly(i**2 + 1, i, modulus=11)

In [49]: gb[10].as_expr().as_poly(modulus=11).ground_roots()
Out[49]: {}

In [51]: gb[10].as_expr().as_poly(modulus=5).ground_roots()
Out[51]: {-2: 1, 2: 1}
```
Depending on what you want a Groebner basis might be slow though. Alternatively after substituting values for all variables you can compute the gcd of the remaining polynomials and find its roots with `ground_roots`.

**Comment by oscarbenjamin** [comment]
One thing I just realised that I hadn't noticed before is that SymPy does not have an implementation of factorisation of multivariate polynomials over `GF(p)`:
```python
In [54]: factor(x + y, modulus=11)
---------------------------------------------------------------------------
NotImplementedError: multivariate polynomials over finite fields
```
Currently python-flint does not have multivariate polynomials over `GF(p)` but multivariate polynomials were very recently added for `ZZ` and `QQ` so I expect that the `nmod_mpoly` and `fmpz_mod_mpoly` types will be added soon and those will undoubtedly provide factorisation.

**Comment by oscarbenjamin** [comment]
I have a tentative partial fix for the comparison issue described in the OP in gh-26795. It makes sympify reject GF elements. That is a breaking change that would have almost certainly broken your code because e.g.:
```python
In [3]: x = symbols('x')

In [4]: x + GF(3)(2)
---------------------------------------------------------------------------
TypeError
```
I describe this as only a partial fix because it still needs a change in python-flint if the expectation is that `nmod` and `Integer` can compare equal. It fixes things on sympy's end but not python-flint's.

It would have been better if this was always the behaviour to avoid confusion. I am not sure if it is worth making the breaking change right now though. It could perhaps be added to a list of candidate breaking changes for a hypothetical future SymPy 2.0. Then again it is possible that the changes in SymPy 1.13.0 would mean that this kind of mixed GF/Expr usage is already broken. I imagine that gh-26795 would be noticed by more people than the changes referred to here though.

The fact that things like this worked until now is just because `sympify` will convert anything that defines `__int__` and `GF(p)` elements define `__int__`. It was not intentional that this would happen because usually domain elements are instances of `CantSympify` which intentionally prevents sympification. I think it was just an oversight that `ModularInteger` did not subclass `CantSympify`.

I also found one place in the sympy codebase that depended on mixed GF/Expr arithmetic in `elliptic_curve.py`.

**Comment by J08nY** [comment]
I am in favor of disallowing this mixing if it is not intended. It is quite natural to expect to be able to use and manipulate finite field elements inside expressions (as one does with integers, rationals or reals).

Btw is it even possible to do that in sympy? Work with expressions over `GF(p)`?

**Comment by oscarbenjamin** [comment]
> Btw is it even possible to do that in sympy? Work with expressions over `GF(p)`?

Yes, if you use Poly or if you work with the domain types directly or if you restrict what operations are used. No if you want it be like SymPy's ordinary symbolic expressions except somehow knowing that you intended positive characteristic.

You can also use Poly functions with the modulus argument:
```python
In [6]: factor(x**2 + 1, modulus=2)
Out[6]: 
       2
(x + 1) 
```
You must avoid division though because it does not have the same meaning (the linked PR would make the examples below both errors):
```python
In [3]: e = GF(5)(2)

In [4]: x + 1/e
Out[4]: x + 3

In [5]: x + S.One/e
Out[5]: x + 1/2
```
On the other hand:
```python
In [28]: x + invert(2, 5)
Out[28]: x + 3
```
You can also invert a polynomial modulo another polynomial in a given positive characteristic:
```python
In [27]: invert(x**2 + 1, x - 1, modulus=5)
Out[27]: -2
```
To use rational functions over GF(p) you can do e.g.:
```python
In [7]: F = GF(2).frac_field(x, y)

In [8]: F.gens[0]**2 / F.gens[1]
Out[8]: x**2/(y)

In [9]: F.gens[0]**2 / F.gens[1] + 2
Out[9]: x**2/(y)
```
The difference is that with Poly or in the domain system you specify what ring you want and so you can specify the characteristic. With ordinary SymPy expressions there is no notion of characteristic and the field is always the field of complex numbers (even if you only work in a subset of the complex numbers). Even the symbolic `Mod` function is defined as a sawtooth function on the real numbers rather than meaning modulo on the sense of finite fields.

**Comment by J08nY** [comment]
Thanks!! This clears up a bunch of things. An explanation like that would be great for the documentation as searching for "sympy modular arithmetic" or "sympy finite fields" leads one to the polys/domains documentation that is not as clear on what you just described.

**Comment by oscarbenjamin** [comment]
Yes, it would be good to have that in the docs. We could have something like a "modular arithmetic and finite fields in sympy" guide.

**Comment by J08nY** [comment]
Great. Thanks for the effort here!
Also, sorry if I came across a bit bitter, half my testsuite breaking by an update was a sad state of affairs but everything is now even better than before.

**Comment by oscarbenjamin** [comment]
No problem and thanks for reporting.

> everything is now even better than before.

Excellent!

Out of interest did you end up with something similar to the above (https://github.com/sympy/sympy/issues/26791#issuecomment-2222823642)?

I notice now that Mathematica has a FiniteFieldElement type which allows using finite fields as part of symbolic expressions. Maybe SymPy could do something similar:

https://reference.wolfram.com/language/ref/FiniteFieldElement.html

**Comment by J08nY** [comment]
> Out of interest did you end up with something similar to the above ([#26791 (comment)](https://github.com/sympy/sympy/issues/26791#issuecomment-2222823642))?

Sort of, I ditched my previous approach of stuffing GFs into an expr and went with stuffing the integer parts in there, then the `as_numer_denom` helped and I was able to get the solutions I needed.

**Comment by oscarbenjamin** [comment]
Well I'm glad it's working now and sorry about the breakage.
